library/src/main/java/com/datatorrent/lib/state/managed/AbstractManagedStateImpl.java

```diff
@@ -0,0 +1,487 @@
+      buckets[bucketIdx] = bucket;
+    }
+    return  bucketIdx;
+  }
+
+  protected int getBucketIdx(long bucketKey)
+  {
+    return (int)(bucketKey % numBuckets);
+  }
+
+  Bucket getBucket(long groupId)
+  {
+    return buckets[getBucketIdx(groupId)];
+  }
+
+  protected Bucket newBucket(long groupId)
+  {
+    return new Bucket.DefaultBucket(groupId, this);
+  }
+
+  @Override
+  public void endWindow()
+  {
+    timeBucketAssigner.endWindow();
+    if (!replay && windowCount == incrementalCheckpointWindowCount) {
+      checkpointDifference();
+      windowCount = 0;
+    }
+  }
+
+  protected void checkpointDifference()
+  {
+    Map<Long, Map<Slice, Bucket.BucketedValue>> flashData = Maps.newHashMap();
+
+    for (Bucket bucket : buckets) {
+      if (bucket != null) {
+        Map<Slice, Bucket.BucketedValue> flashDataForBucket = bucket.checkpoint(windowId);
+        if (!flashDataForBucket.isEmpty()) {
+          flashData.put(bucket.getBucketKey(), flashDataForBucket);
+        }
+      }
+    }
+    if (!flashData.isEmpty()) {
+      try {
+        dataManager.save(flashData, operatorId, windowId);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  @Override
+  public void checkpointed(long l)
+  {
+  }
+
+  @Override
+  public void committed(long l)
+  {
+    synchronized (commitLock) {
+      try {
+        for (Bucket bucket : buckets) {
+          if (bucket != null) {
+            bucket.committed(l);
+          }
+        }
+        dataManager.committed(operatorId, l);
+      } catch (IOException e) {
+        throw new RuntimeException("committing " + l, e);
+      }
+    }
+  }
+
+  @Override
+  public void teardown()
+  {
+    dataManager.teardown();
+    timeBucketAssigner.teardown();
+    for (Bucket bucket : buckets) {
+      if (bucket != null) {
+        bucket.teardown();
+      }
+    }
+    stateTracker.teardown();
+  }
+
+  @Override
+  public void setCacheSize(long bytes)
+  {
+    maxCacheSize = bytes;
+  }
+
+  public long getCacheSize()
+  {
+    return maxCacheSize;
+  }
+
+  public void setFileAccess(@NotNull FileAccess fileAccess)
+  {
+    this.fileAccess = Preconditions.checkNotNull(fileAccess);
+  }
+
+  public void setKeyComparator(@NotNull Comparator<Slice> keyComparator)
+  {
+    this.keyComparator = Preconditions.checkNotNull(keyComparator);
+  }
+
+  @Override
+  public FileAccess getFileAccess()
+  {
+    return fileAccess;
+  }
+
+  public void setTimeBucketAssigner(@NotNull TimeBucketAssigner timeBucketAssigner)
+  {
+    this.timeBucketAssigner = Preconditions.checkNotNull(timeBucketAssigner);
+  }
+
+  public TimeBucketAssigner getTimeBucketAssigner()
+  {
+    return timeBucketAssigner;
+  }
+
+  @Override
+  public Comparator<Slice> getKeyComparator()
+  {
+    return keyComparator;
+  }
+
+  @Override
+  public BucketsMetaDataManager getBucketsMetaDataManager()
+  {
+    return bucketsMetaDataManager;
+  }
+
+  public int getNumReaders()
```

Comment:
Haven't completed the documentation yet. There are multiple threads that can load different buckets from disk. This configures the number of threads in the pool. This is why synchronizing on a bucket in the async call so that fetching of a single bucket cannot is done by a single thread.